### ISSUE TITLE
thread_pool_executor deadlock

### ISSUE THREAD

**Issue by lordvictory** [David-Haim/concurrencpp#61]
Fascinated with your toolkit and while experimenting may have discovered a possible executor deadlock bug.  It occurs after submitting multiple result sets and awaiting completion via when_all. The same issue had also been observed using when_any as well.

I've managed to isolate a relatively small test case which I'll attach, that uses when_all.

In my test case I call submit() in a 2 level nested fashion, having passed the executor in as an argument.

At the second level the code may be conditionally compiled to substitute using std_thread instead of submit().  When using submit() a deadlock is observed. When using std::thread no deadlock is observed.

Am using MSVC 19.29.30136 for x64 (as reported from command line using cl --version).

Dev tools environment is:

WIndows 10 hosted VM configured for 2 CPU's.

Microsoft Visual Studio Enterprise 2019
Version 16.11.5
VisualStudio.16.Release/16.11.5+31729.503

Installed Version: Enterprise

Visual C++ 2019   00435-60000-00000-AA788
Microsoft Visual C++ 2019

Here is a test case below. Looks like the inserting into the forum here screwed up the source code a bit, particularly greater-than/less-than signs. But after reading the code you should be able to get the idea. Hope this helps.

#include "iostream"
#include "chrono"
#include "string"
#include "ctime"

#include "concurrencpp/concurrencpp.h"

///////////////////////////////////////////////////////////////////////////////////////
int   dummy_task2(std::shared_ptr<concurrencpp::thread_pool_executor> background_executor, int instance)
{
    std::cerr << "+dummy_task2(): instance=" << instance << ", tid=" << std::this_thread::get_id() << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds((instance+1) * 250));
    std::cerr << "-dummy_task2(): instance=" << instance << ", tid=" << std::this_thread::get_id() << std::endl;
    return 0;
}

///////////////////////////////////////////////////////////////////////////////////////
concurrencpp::result<int>   level2_dispatch(std::shared_ptr<concurrencpp::thread_pool_executor> background_executor)
{
    std::vector<concurrencpp::result<int>>  results;

    std::cerr << "+level2_dispatch(), tid=" << std::this_thread::get_id() << std::endl;

    for (int i = 0; i < 1; i++)
    {
        // start a thread for each...
//#define USE_STD_THREAD
#if defined(USE_STD_THREAD)
        concurrencpp::result_promise<int>  promise;
        auto res = promise.get_result();
        std::thread  my_3_party_executor([promise = std::move(promise), background_executor, i]() mutable
        {
            promise.set_result(dummy_task2(background_executor, i));
        });
        my_3_party_executor.detach();
        results.emplace_back(std::move(res));
#else
        results.emplace_back(background_executor->submit(dummy_task2, background_executor, i));
#endif
    }

    auto all_consumed = concurrencpp::when_all(std::begin(results), std::end(results));
    all_consumed.get();

    std::cerr << "-level2_dispatch(), tid=" << std::this_thread::get_id() << std::endl;
    co_return 0;
}

///////////////////////////////////////////////////////////////////////////////////////
int   dummy_task1(std::shared_ptr<concurrencpp::thread_pool_executor> background_executor, int instance)
{
    std::cerr << "+dummy_task1(): instance=" << instance << ", tid=" << std::this_thread::get_id() << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds((instance+1) * 250));
    try
    {
        level2_dispatch(background_executor).get();
    }
    catch (const std::exception& e)
    {
        std::cerr << "@main(): EXCEPTION: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }
    std::cerr << "-dummy_task1(): instance=" << instance << ", tid=" << std::this_thread::get_id() << std::endl;
    return 0;
}

///////////////////////////////////////////////////////////////////////////////////////
concurrencpp::result<int>   level1_dispatch(std::shared_ptr<concurrencpp::thread_pool_executor> background_executor)
{
    std::vector<concurrencpp::result<int>>  results;

    std::cerr << "+level1_dispatch(), tid=" << std::this_thread::get_id() << std::endl;
    for (int i = 0; i < 1; i++)
    {
        // start a thread for each...
        results.emplace_back(background_executor->submit(dummy_task1, background_executor, i));
    }

    auto all_consumed = concurrencpp::when_all(std::begin(results), std::end(results));
    all_consumed.get();

    std::cerr << "-level1_dispatch(), tid=" << std::this_thread::get_id() << std::endl;
    co_return 0;
}

///////////////////////////////////////////////////////////////////////////////////////
int     main(int argc, char *argv[])
{
    concurrencpp::runtime   runtime;

    std::cerr << "+main(): tid=" << std::this_thread::get_id() << std::endl;
    try
    {
        level1_dispatch(runtime.background_executor()).get();
    }
    catch (const std::exception& e)
    {
        std::cerr << "@main(): EXCEPTION: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }
    std::cerr << "-main(): tid=" << std::this_thread::get_id() << std::endl;
}

**Comment by David-Haim** [comment]
I haven't really debugged the code (it needs fixing in order to compile it) but the problem is (I think) clear: you block waiting for `when_all` to finish instead of `co_await`ing it.

As the documentation says :

 Waiting for a result to finish is a blocking operation (in the case the asynchronous result is not ready), and will suspend the entire thread of execution waiting for the asynchronous result to become available. Waiting operations are generally discouraged and only allowed in root-level tasks or in contexts which allow it, like blocking the main thread waiting for the rest of the application to finish gracefully, or using concurrencpp::blocking_executor or concurrencpp::thread_executor.

(Do note that for that specific example you posted, even `background_executor` is not enough)

So what happens is this : you submit more and more tasks, creating a hierarchy of dependencies between scheduled tasks, but you also block the threadpool itself waiting for itself to finish executing every task. after some load, it will break.

The solution is simple: you need to `co_await` `when_all` rather than to `get` it:

before:
```cpp
    auto all_consumed = concurrencpp::when_all(std::begin(results), std::end(results));
    all_consumed.get();
```

after:
```cpp
    auto all_consumed =  co_await concurrencpp::when_all(std::begin(results), std::end(results));
```

Does the problem go away after this change?

**Comment by David-Haim** [comment]
Addendum:
Here's what I did:

* I got rid of `get` entirely - except the main thread waiting for the entire application to finish
* I changed `dummy_task1` to be a parallel coroutine, so `level1_dispatch` doesn't have to schedule a task that already return a `result` (yielding `result<result<int>`).
* made the application schedule 1000 * 1000 total `dummy_task2`
* lowered the sleeping time. you can revert it to be 250ms , the example will take a lot of time to finish, but it will finish eventually (within hours if not days).

Note that blocking inside threadpool-threads is a common problem among threadpools, this is why `co_await` works so well with this design - you can consume asynchronous results without interfering with the threadpool inner mechanism.

Interestingly, the Java Fork-Join pool "knows" to handle blocking by letting the developer notify the threadpool the task is about to block (using the  ForkJoinPool.ManagedBlocker interface), then the threadpool knows to inject new threads to prevent starvation. I will say it's highly inefficient and inferior to coroutines, which don't have to block (they're suspended instead).

[main.txt](https://github.com/David-Haim/concurrencpp/files/7491029/main.txt)

**Comment by lordvictory** [comment]
Thank you so much for taking the time to clarify proper usage.  Once I corrected my errors everything ran as expected.

I've been doing "old fashioned" c++ concurrent programming (i.e. unstructured) for several decades now..very messy and hard to maintain.  But, the new additions to c++ 20 for "modern" concurrent programming (i.e. structured) are  amazingly powerful.  Yet, without a toolkit such as concurrencpp are relatively useless to most programmers. Your toolkit provides an efficient abstraction layer that is needed to make sense out of what the language implementors provided.

Again, thank-you so much for your hard work, and for taking the time to explain how to best utilize your toolkit.
